:sparkles: ClusterClass patches: Allow matching all MachineDeploymentClasses

valaparthvi opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:

This PR adds a new feature to allow matching all "*" MachineDeploymentClasses.
1. Match all if MachineDeploymentClass.Names is a "*".
```yaml
matchResources:
  machineDeploymentClass:
    names : [*]
```
2. Match some if explicitly mentioned.
```yaml
matchResources:
  machineDeploymentClass:
    names: [something, something, something]
```
3. Match some if glob matched.
```yaml
matchResources:
  machineDeploymentClass:
    names: [something-*]
```

3. Match none if no machineDeploymentClass  is set (should produce a webhook failure)
```yaml
matchResources:
  machineDeploymentClass: {}
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5648

killianmuldoon reviewed (commented):
Overall I think this might be useful, but is there a cleaner way to do it in the API? Just not a fan of using regex-lookalike syntax if we're not supporting any real pattern matching. Would it be reasonable to have the empty selector act as a match all? i.e. https://github.com/kubernetes/apimachinery/blob/afc5e00a762a432aa7d988d711dab6dcf3427585/pkg/apis/meta/v1/helpers.go#L38

sbueringer commented:
I think using `*` which is already established in Kubernetes for ClusterRoles/Roles is better than using something like empty or no selector. It's way more explicit that way

killianmuldoon commented:
> How would an empty selector look like for the names slice that we have today?
```
        matchResources:
          machineDeploymentClass: {}
```

That's my understanding of how it normally works in Kubernetes with the label selector.  It would require enabling it (and maybe even setting it) in the webhook. 

I think it's more consistent to stick to how selectors work in Kubernetes rather than RBAC, despite it being implicit. Definitely not a blocker for implementing it as is with the '*' notation, but it does depart from convention IMO.

killianmuldoon commented:
/ok-to-test

valaparthvi commented:
@killianmuldoon @sbueringer So, what have you decided? Are we going with K8s way or RBAC way?
The K8s way does not seem very intuitive to me, or if that is the way it happens generally, it might be intuitive for others. RBAC on the other hand is quite simple.

sbueringer commented:
Thought about it a bit more. I think Killian is right. It makes sense to stay consistent with Kubernetes selectors.

This would give us the following semantics:

Match specific MD classes
```
        matchResources:
          machineDeploymentClass:
            names:
            - default-worker
```

Match all MD classes
```
        matchResources:
          machineDeploymentClass: {}
```

Match none MD classes
```
        matchResources:
```

We should not set the selector in the defaulting webhook. The idea in the ClusterClass API is that folks should explicitly decide to select InfraCluster, CP, MD classes, ... . If we would default to `{}` folks have to set some not-matching MD classes to still be able to not match MD classes.

(I think it comes down to in Kubernetes it's often default: "match all", and in our cases the default is "match none")

Independent of that, I absolutely don't like what they did there it is way to implicit for my taste.

sbueringer commented on internal/controllers/topology/cluster/patches/inline/json_patch_generator.go:
```diff
@@ -167,6 +167,10 @@ func matchesSelector(req *runtimehooksv1.GeneratePatchesRequestItem, templateVar
 
 			// If the builtin variable could be read.
 			if err == nil {
+				// If templateMDClassJSON is defined, but no values are passed, consider it matches all
+				if selector.MatchResources.MachineDeploymentClass.Names != nil && len(selector.MatchResources.MachineDeploymentClass.Names) == 0 {
+					return true
+				}
```
```suggestion
				// If templateMDClassJSON is defined and the selector is set to an empty object, consider it matches all.
				if len(selector.MatchResources.MachineDeploymentClass.Names) == 0 {
					return true
				}
```
nit: just a small simplification + go doc change

I think we don't really care about if the array is nil or not nil. We just want to check if it's empy in some way.

sbueringer replied:
Please add the following comment to the API type field ~ l.394 in clusterclass_types.go
```go
	// Note: If set to an empty object it will match all MachineDeploymentClasses.
```

sbueringer commented on internal/controllers/topology/cluster/patches/inline/json_patch_generator_test.go:
```diff
@@ -617,6 +617,71 @@ func TestMatchesSelector(t *testing.T) {
+			match: true,
+		},
+		{
+			name: "Match none MD BootstrapTemplate",
```
I think this case should match as well (which it should after the change I proposed above).

The case which shouldn't match is if `MachineDeploymentClass` is set to nil. Can you please add an additional test case for it?

valaparthvi commented:
@sbueringer I can't seem to run unit tests for patch_test.go locally, neither from my branch, and nor from main branch. Do I need to run some setup? I have a kind cluster running.
<details>
<summary>Here is the output from when I run tests.</summary>

```console
/usr/local/go/bin/go tool test2json -t /tmp/GoLand/___TestPatchHelper_in_sigs_k8s_io_cluster_api_util_patch.test -test.v -test.paniconexit0 -test.run ^\QTestPatchHelper\E$
E0727 18:57:10.811564  125075 logr.go:265] controller-runtime/test-env "msg"="unable to start the controlplane" "error"="fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory"  "tries"=0
E0727 18:57:10.812226  125075 logr.go:265] controller-runtime/test-env "msg"="unable to start the controlplane" "error"="fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory"  "tries"=1
E0727 18:57:10.812763  125075 logr.go:265] controller-runtime/test-env "msg"="unable to start the controlplane" "error"="fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory"  "tries"=2
E0727 18:57:10.813390  125075 logr.go:265] controller-runtime/test-env "msg"="unable to start the controlplane" "error"="fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory"  "tries"=3
E0727 18:57:10.813975  125075 logr.go:265] controller-runtime/test-env "msg"="unable to start the controlplane" "error"="fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory"  "tries"=4
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x14f55af]

goroutine 1 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/testing/controlplane.(*APIServer).Stop(0x14)
	/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api/vendor/sigs.k8s.io/controller-runtime/pkg/internal/testing/controlplane/apiserver.go:424 +0x8f
sigs.k8s.io/controller-runtime/pkg/internal/testing/controlplane.(*ControlPlane).Stop(0xc0005af180)
	/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api/vendor/sigs.k8s.io/controller-runtime/pkg/internal/testing/controlplane/plane.go:97 +0x3d
sigs.k8s.io/controller-runtime/pkg/envtest.(*Environment).Stop(0xc0005af180)
	/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api/vendor/sigs.k8s.io/controller-runtime/pkg/envtest/server.go:194 +0x114
sigs.k8s.io/cluster-api/internal/test/envtest.newEnvironment({0x0, 0x0, 0xc000598600})
	/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api/internal/test/envtest/environment.go:209 +0x22e5
sigs.k8s.io/cluster-api/internal/test/envtest.Run({0x1de86b0, 0xc0004ccf40}, {0xc000184000, {0x0, 0x0, 0x0}, 0x0, 0x0, 0x1c2d2a8, {0x0, ...}})
	/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api/internal/test/envtest/environment.go:112 +0x66
sigs.k8s.io/cluster-api/util/patch.TestMain(0xc0000001a0)
	/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api/util/patch/suite_test.go:39 +0x66
main.main()
	_testmain.go:51 +0x165
```
</details>

killianmuldoon commented:
The issue seems to be your env test setup -  do all the tests run when you run `make test` ? You may need to set the `
`KUBEBUILDER_ASSETS` env variable

valaparthvi commented:
> The issue seems to be your env test setup - do all the tests run when you run `make test` ? 
You may need to set the `KUBEBUILDER_ASSETS` env variable

I was actually running the tests directly from my GoLand editor, and I didn't realize I had to set the envvar, but it seems like `make test` passed on both the branches.

I also noticed pull-cluster-api-verify-main is failing because I failed to update config/crd/bases/cluster.x-k8s.io_clusterclasses.yaml. Is there a Make target that I should run to update this yaml or will this be a manual change?

killianmuldoon commented:
The CR definitions are created using `make generate` and its sub-targets. `make generate` takes a long time to run, so you might want to just run `make generate-manifests` and maybe `make generate-openapi` 

Note - it's not a good idea to run `make verify-gen` locally as it runs the full `make generate` if you're in doubt, you can just run the full job (takes about 20 minutes for me locally, but heavily depends on setup and OS.

ykakarap commented on api/v1beta1/clusterclass_types.go:
```diff
@@ -391,6 +391,7 @@ type PatchSelectorMatch struct {
 
 	// MachineDeploymentClass selects templates referenced in specific MachineDeploymentClasses in
 	// .spec.workers.machineDeployments.
+	// Note: If set to an empty object it will match all MachineDeploymentClasses.
```
How about also add a comment on `.PatchSelectorMatchMachineDeploymentClass.Names` that reads: `Note: If the list is empty then it sill match all MachineDeploymentClass`? 
Because, I think 
```
        matchResources:
          controlPlane: true
          machineDeploymentClass:
                name: []
```
will also pass the implemented logic and it will match all MachineDeploymentClasses. 

This is not blocking for the PR.

valaparthvi replied:
You're right. Apart from the note, I have also added a unit test just to be on the safe side.

ykakarap reviewed (commented):
LGTM pending squash.

valaparthvi commented:
> The CR definitions are created using `make generate` and its sub-targets. `make generate` takes a long time to run, so you might want to just run `make generate-manifests` and maybe `make generate-openapi`
> 
> Note - it's not a good idea to run `make verify-gen` locally as it runs the full `make generate` if you're in doubt, you can just run the full job (takes about 20 minutes for me locally, but heavily depends on setup and OS.

Thank you for this! And I think it's `make generate-go-openapi`, and not `make generate-openapi`, I couldn't find that target.

valaparthvi commented:
I am not sure how to fix pull-cluster-api-verify-main job. I've run a different make targets locally, but I seem to have reached nowhere. I've run `make generate`, which modifies a few files, none of which contains the file in question(api/v1beta1/zz_generated.openapi.go). I've also run `make generate-go-openapi` which seems to be doing nothing.

<details>
<summary> O/P from `make generate`</summary>

```console
git status
On branch 5648-allow-matching-all-mdc
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	renamed:    api/v1alpha3/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/api/v1alpha3/zz_generated.conversion.go
	renamed:    api/v1alpha4/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/api/v1alpha4/zz_generated.conversion.go
	renamed:    bootstrap/kubeadm/api/v1alpha3/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/bootstrap/kubeadm/api/v1alpha3/zz_generated.conversion.go
	renamed:    bootstrap/kubeadm/api/v1alpha4/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/bootstrap/kubeadm/api/v1alpha4/zz_generated.conversion.go
	renamed:    bootstrap/kubeadm/types/upstreamv1beta1/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/bootstrap/kubeadm/types/upstreamv1beta1/zz_generated.conversion.go
	renamed:    bootstrap/kubeadm/types/upstreamv1beta2/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/bootstrap/kubeadm/types/upstreamv1beta2/zz_generated.conversion.go
	renamed:    bootstrap/kubeadm/types/upstreamv1beta3/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/bootstrap/kubeadm/types/upstreamv1beta3/zz_generated.conversion.go
	renamed:    controlplane/kubeadm/api/v1alpha3/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/controlplane/kubeadm/api/v1alpha3/zz_generated.conversion.go
	renamed:    controlplane/kubeadm/api/v1alpha4/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/controlplane/kubeadm/api/v1alpha4/zz_generated.conversion.go
	renamed:    exp/addons/api/v1alpha3/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/exp/addons/api/v1alpha3/zz_generated.conversion.go
	renamed:    exp/addons/api/v1alpha4/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/exp/addons/api/v1alpha4/zz_generated.conversion.go
	renamed:    exp/api/v1alpha3/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/exp/api/v1alpha3/zz_generated.conversion.go
	renamed:    exp/api/v1alpha4/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/exp/api/v1alpha4/zz_generated.conversion.go
	renamed:    internal/runtime/test/v1alpha1/zz_generated.conversion.go -> github.com/kubernetes-sigs/cluster-api/internal/runtime/test/v1alpha1/zz_generated.conversion.go
	new file:   test/infrastructure/docker/github.com/kubernetes-sigs/cluster-api/test/infrastructure/docker/api/v1alpha3/zz_generated.conversion.go
	new file:   test/infrastructure/docker/github.com/kubernetes-sigs/cluster-api/test/infrastructure/docker/api/v1alpha4/zz_generated.conversion.go
	new file:   test/infrastructure/docker/github.com/kubernetes-sigs/cluster-api/test/infrastructure/docker/exp/api/v1alpha3/zz_generated.conversion.go
	new file:   test/infrastructure/docker/github.com/kubernetes-sigs/cluster-api/test/infrastructure/docker/exp/api/v1alpha4/zz_generated.conversion.go
```

</details>

Can someone help me fix this?

ykakarap commented on internal/controllers/topology/cluster/patches/inline/json_patch_generator_test.go:
```diff
@@ -617,6 +617,100 @@ func TestMatchesSelector(t *testing.T) {
+			match: true,
+		},
+		{
+			name: "Match all MD BootstrapTemplate",
```
```suggestion
			name: "Match all MD BootstrapTemplate for an empty .PatchSelectorMatchMachineDeploymentClass object",
```

ykakarap reviewed (commented):
Can you share the output of `make generate-go-openapi`?

I was able to run `make generate-go-openapi` on this PR and it generated the required change to fix the issue in `pull-cluster-api-verify-main` job. So, I don't the issue is with any of the changes that are part of this PR.

It is weird that when you run `make generate` the generated files are getting renamed 🤔 . Is your project clones outside the GOPATH (although that should not cause any problems as there is logic in the make file to take care of it: https://github.com/valaparthvi/cluster-api/blob/e0bdc2455bef5e537888f10c45149f0df6cede72/Makefile#L65-L71)?

killianmuldoon commented:
@valaparthvi When I run `make generate-go-openapi` locally I get changes in ` api/v1beta1/zz_generated.openapi.go`. Could you share some information about the environment in which you're running the command? e.g. cluster api most recent commit (before your changes),  os type and kernel version, and the directory in which you're running the make command. 

i.e. the output of:
1.  `git log --oneline | head -n 5` 
2. `go env | grep 'GOOS\|GOARCH\|GOVERSION\|GOPATH'`
3. The output of `pwd` in the directory you're running the make commands.

Note the above will likely only work for Mac / Linux. If you're in a Windows environment you may need to find a different path to providing that information.

valaparthvi commented:
I ran `make generate-go-openapi` but it did not make any change to fix the failure.
<details>
<summary>Here's the o/p from when I ran it</summary>

```console
make generate-go-openapi                   
GOBIN=/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api/hack/tools/bin ./scripts/go_install.sh k8s.io/kube-openapi/cmd/openapi-gen openapi-gen 5e7f5fd
make[1]: Entering directory '/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api/tmp'
make[1]: *** No rule to make target 'clean-generated-openapi-definitions'.  Stop.
make[1]: Leaving directory '/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api/tmp'
** Generating openapi schema for types in ./api/v1beta1 **
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,ClusterClassPatch,Definitions
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,ClusterClassSpec,Patches
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,ClusterClassSpec,Variables
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,JSONSchemaProps,Enum
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,JSONSchemaProps,Required
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,MachineDeploymentVariables,Overrides
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,MachineHealthCheckClass,UnhealthyConditions
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,MachineHealthCheckSpec,UnhealthyConditions
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,MachineHealthCheckStatus,Targets
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,NetworkRanges,CIDRBlocks
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,PatchDefinition,JSONPatches
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,PatchSelectorMatchMachineDeploymentClass,Names
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,Topology,Variables
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,WorkersClass,MachineDeployments
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/api/v1beta1,WorkersTopology,MachineDeployments
make[1]: Entering directory '/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api/tmp'
make[1]: *** No rule to make target 'clean-generated-openapi-definitions'.  Stop.
make[1]: Leaving directory '/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api/tmp'
** Generating openapi schema for types in ./exp/runtime/hooks/api/v1alpha1 **
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1,GeneratePatchesRequest,Items
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1,GeneratePatchesRequest,Variables
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1,GeneratePatchesRequestItem,Variables
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1,GeneratePatchesResponse,Items
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1,GeneratePatchesResponseItem,Patch
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1,ValidateTopologyRequest,Items
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1,ValidateTopologyRequest,Variables
API rule violation: list_type_missing,sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1,ValidateTopologyRequestItem,Variables
```

</details>

```console
$ git status                                     
On branch 5648-allow-matching-all-mdc
nothing to commit, working tree clean
```

>Is your project clones outside the GOPATH ?

No. It is within the GOPATH.

```console
$ git log --oneline | head -n 5
0142e0b18 Update internal/controllers/topology/cluster/patches/inline/json_patch_generator_test.go
e0bdc2455 Fix crd API docs
25b6cd93b ykakarap's review: add unit test to match .PatchSelectorMachineDeploymentClass.Names to an empty object, and a note about it
206e507c2 requested changes
989c45e10 Update internal/controllers/topology/cluster/patches/inline/json_patch_generator.go
```

```console
$ go env | grep 'GOOS\|GOARCH\|GOVERSION\|GOPATH'
GOARCH="amd64"
GOOS="linux"
GOPATH="/home/pvala/go"
GOVERSION="go1.17.12"
```

```console
$ pwd
/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api

```

killianmuldoon commented:
I'm really not sure what the base issue is here at this point 🤔, and I don't know what the next steps to look for a solution are. Just as a sanity check can you rebase on main and try to run the make generate targets again? Once that's done I think we should focus on the substance of the PR. We should be able to fix up the generated OpenAPI stuff on your branch before it's time to merge.

valaparthvi commented:
Same result with the rebase.
<details>
<summary>Here's the output from `git status` after I ran `make generate`</summary>

```console
git status                         
On branch 5648-allow-matching-all-mdc
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	deleted:    api/v1alpha3/zz_generated.conversion.go
	deleted:    api/v1alpha4/zz_generated.conversion.go
	deleted:    bootstrap/kubeadm/api/v1alpha3/zz_generated.conversion.go
	deleted:    bootstrap/kubeadm/api/v1alpha4/zz_generated.conversion.go
	deleted:    bootstrap/kubeadm/types/upstreamv1beta1/zz_generated.conversion.go
	deleted:    bootstrap/kubeadm/types/upstreamv1beta2/zz_generated.conversion.go
	deleted:    bootstrap/kubeadm/types/upstreamv1beta3/zz_generated.conversion.go
	deleted:    controlplane/kubeadm/api/v1alpha3/zz_generated.conversion.go
	deleted:    controlplane/kubeadm/api/v1alpha4/zz_generated.conversion.go
	deleted:    exp/addons/api/v1alpha3/zz_generated.conversion.go
	deleted:    exp/addons/api/v1alpha4/zz_generated.conversion.go
	deleted:    exp/api/v1alpha3/zz_generated.conversion.go
	deleted:    exp/api/v1alpha4/zz_generated.conversion.go
	deleted:    internal/runtime/test/v1alpha1/zz_generated.conversion.go

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	github.com/
	test/infrastructure/docker/github.com/
```

</details>

I am not sure about the 2 new folders added.

killianmuldoon commented:
Can you show what the rest of the path is under those newly created github directories?

chrischdi commented:
Could you try to checkout/move the repository to `/home/pvala/go/src/sigs.k8s.io/cluster-api` instead of `/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api`? 

Maybe this fixes the generate 🤔

valaparthvi commented:
> Could you try to checkout/move the repository to `/home/pvala/go/src/sigs.k8s.io/cluster-api` instead of `/home/pvala/go/src/github.com/kubernetes-sigs/cluster-api`?
> 
> Maybe this fixes the generate thinking

Surprisingly, that did it.

ykakarap commented:
There is code in Makefile that should make this work even if the repo is cloned outside the `GOPATH`. I wonder why it did not work. 
Ref: https://github.com/kubernetes-sigs/cluster-api/blob/2fc48a831e444981cbdc52c0d41b19571235e90b/Makefile#L65-L71

killianmuldoon commented:
>There is code in Makefile that should make this work even if the repo is cloned outside the GOPATH. I wonder why it did not work.

Yeah - I tried this locally using a similar path and it worked fine 🤔

killianmuldoon commented:
This is shaping up nicely! There's still a change needed in the webhook to enable creating ClusterClasses with the empty selector (and a test to show it's working)

The relevant code is here:
https://github.com/kubernetes-sigs/cluster-api/blob/998bc05c71d588109a426ec4019bf73612659c0a/internal/webhooks/patch_validation.go#L158-L161

My preference would be for the webhook to reject 

```go
        matchResources:
          machineDeploymentClass:
                name: []
```
And accept
```go
        matchResources:
          machineDeploymentClass: {}
```

I think that cuts down on the complexity of the API. The empty name list is another added layer of confusion IMO.

ykakarap reviewed (commented):
>The empty name list is another added layer of confusion IMO.

Agree that it can be confusing. 

I guess the only way to check for this in the webhook would be to fail the validation if `selector.MatchResources.MachineDeploymentClass.Names != nil && len(selector.MatchResources.MachineDeploymentClass.Names) == 0`.

@valaparthvi Can you please make sure to add a comment explaining why we are doing this particular check with the example showing when it will fail the validation?

valaparthvi commented:
@killianmuldoon @ykakarap Just to be clear, you both are suggesting addition of one more condition check, right? The existing checks and tests should stay as is?

killianmuldoon commented:
I think we should add some tests to the webhook to cover the following cases:
1. Matching all MDs works to select multiple MachineDeployments when the selector is empty i.e. {}
2. Setting the selector to {Names:[]} is rejected - returns an error

valaparthvi commented:
> I think we should add some tests to the webhook to cover the following cases:
> 
>     1. Matching all MDs works to select multiple MachineDeployments when the selector is empty i.e. {}
> 
>     2. Setting the selector to {Names:[]} is rejected - returns an error

Right. I postponed writing the tests until one of you approved the code changes.

ykakarap commented on api/v1beta1/clusterclass_types.go:
```diff
@@ -399,6 +400,7 @@ type PatchSelectorMatch struct {
 // in specific MachineDeploymentClasses in .spec.workers.machineDeployments.
 type PatchSelectorMatchMachineDeploymentClass struct {
 	// Names selects templates by class names.
+	// Note: If the list is empty then it will match all MachineDeploymentClass.
```
With the new checks, this comment is no longer applicable. Can you please delete it?

ykakarap commented on internal/controllers/topology/cluster/patches/inline/json_patch_generator_test.go:
```diff
@@ -617,6 +617,100 @@ func TestMatchesSelector(t *testing.T) {
 			match: true,
 		},
+		{
+			name: "Match all MD BootstrapTemplate with .PatchSelectorMatchMachineDeploymentClass.Names set to an empty string array",
```
Similarly, this test case is no longer applicable applicable.

ykakarap commented on internal/webhooks/patch_validation.go:
```diff
@@ -158,7 +158,8 @@ func validateSelectors(selector clusterv1.PatchSelector, class *clusterv1.Cluste
 	if !(selector.MatchResources.InfrastructureCluster || selector.MatchResources.ControlPlane ||
-		(selector.MatchResources.MachineDeploymentClass != nil && len(selector.MatchResources.MachineDeploymentClass.Names) > 0)) {
+		(selector.MatchResources.MachineDeploymentClass != nil && len(selector.MatchResources.MachineDeploymentClass.Names) > 0) ||
+		(selector.MatchResources.MachineDeploymentClass.Names != nil && len(selector.MatchResources.MachineDeploymentClass.Names) == 0)) {
```
```suggestion
		(selector.MatchResources.MachineDeploymentClass != nil && selector.MatchResources.MachineDeploymentClass.Names == nil)) {
```
Isn't it?

ykakarap commented on internal/webhooks/patch_validation.go:
```diff
@@ -197,6 +198,16 @@ func validateSelectors(selector clusterv1.PatchSelector, class *clusterv1.Cluste
 			}
 		}
 	}
+	if selector.MatchResources.MachineDeploymentClass.Names != nil && len(selector.MatchResources.MachineDeploymentClass.Names) == 0 {
```
```suggestion
	if selector.MatchResources.MachineDeploymentClass != nil && selector.MatchResources.MachineDeploymentClass.Names == nil {
```

ykakarap reviewed (commented):
I think you are alreaedy plannin on doing this, so can you please add some tests for the changes to `validateSelectors`.

ykakarap commented on internal/webhooks/patch_validation_test.go:
```diff
@@ -1836,6 +1836,76 @@ func Test_validateSelectors(t *testing.T) {
+				Build(),
+			wantErr: true,
+		},
 		{
```
Can you please add the following tests:
* A test with only `MachineDeploymentClass: &clusterv1.PatchSelectorMatchMachineDeploymentClass{},` and not control plane matching.
* A test where we check for the validation to not pass if it violates the new validation check that was added.

valaparthvi replied:
> A test where we check for the validation to not pass if it violates the new validation check that was added.

I think that is already in place. See [this](https://github.com/kubernetes-sigs/cluster-api/pull/6930/files#diff-235f29a9fd9322b88ccd591b68f30daee2fa9196d92c425700168461cf5e4a5aR1887). Or are you talking the validation where MachineDeploymentClass is nil?

ykakarap replied:
Never mind. You are right. The cases are covered.

ykakarap reviewed (commented):
This is great work. Good job and thank you for working on this.  🙂

killianmuldoon commented on internal/webhooks/patch_validation_test.go:
```diff
@@ -1836,6 +1836,110 @@ func Test_validateSelectors(t *testing.T) {
+			wantErr: false,
+		},
+		{
+			name: "fail if selector targets with an empty MachineDeploymentClass.Name",
```
I think it's okay to include this test, but I don't think it's realistic currently. I don't think there's any way to create an empty `Names` array with the current implementation - It's rendered as nil.

killianmuldoon commented on internal/webhooks/patch_validation.go:
```diff
@@ -183,20 +183,32 @@ func validateSelectors(selector clusterv1.PatchSelector, class *clusterv1.Cluste
-	if selector.MatchResources.MachineDeploymentClass != nil && len(selector.MatchResources.MachineDeploymentClass.Names) > 0 {
-		for _, name := range selector.MatchResources.MachineDeploymentClass.Names {
+	if selector.MatchResources.MachineDeploymentClass != nil {
+		if selector.MatchResources.MachineDeploymentClass.Names == nil {
```
I think we should fail if the MachineDeploymentClass selector is defined (whether or not it's empty) if there are no MachineDeploymentClasses defined. i.e. if `len(class.Spec.Workers.MachineDeployments) == nil`

sbueringer replied:
But isn't that exactly the case we want to support? That `MachineDeploymentClass` is set to an empty object which means:
* selector.MatchResources.MachineDeploymentClass != nil  and
* selector.MatchResources.MachineDeploymentClass.Names == nil

Or am I missing something?

killianmuldoon replied:
Just based on https://github.com/kubernetes-sigs/cluster-api/pull/6930#issuecomment-1220702490 my idea was to reject `MachineDeploymentClass{Names=[]}`, but internally that looks exactly the same as `MachineDeploymentClass{}` no big issue though.

killianmuldoon replied:
What I meant here is we should fail if there is no MachineDeploymentClass defined in the ClusterClass at all i.e. the selector can't ever match anything.

sbueringer replied:
Ah sorry misread your comment. 

You mean PatchSelectorMatchMachineDeploymentClass.Names = [], right? (not `MachineDeploymentClass`)

That sounds fine to me, although I think this case already lands in l.213 given that neither if (l.187) nor  else if (l.196) will match

sbueringer replied:
Okay read the newer version of your comment. Looks like we already fail in that case today, but fine for me to improve the error message for that case.

ykakarap replied:
>I misunderstood the behaviour here this is working as expected.

Just to confirm: Looks like the for loop in l.188 will not trigger if there are no MachineDeploymentClasses are defined in the ClusterClass and it will fall to l.213 and return the correct error. 

Looks like it is working as expected. 👍

killianmuldoon reviewed (commented):
This is looking good - just a couple of cleanups to unit tests and validation and I think this is ready to merge.

sbueringer commented on api/v1beta1/clusterclass_types.go:
```diff
@@ -399,6 +400,7 @@ type PatchSelectorMatch struct {
 // in specific MachineDeploymentClasses in .spec.workers.machineDeployments.
 type PatchSelectorMatchMachineDeploymentClass struct {
 	// Names selects templates by class names.
+	// Note: If the list is empty then it will match all MachineDeploymentClass.
```
Looks like you were reviewing an outdated version of the PR

killianmuldoon replied:
😆 More like I added this comment to the PR weeks ago and never "completed" the review, so this got published with my newer comments.

sbueringer commented:
@valaparthvi @killianmuldoon I've updated the PR description to reflect what we want to achieve. Am I correct?

sbueringer commented:
> ```yaml
> names
> ```

Would be really great if someone can test this with YAML once we have the final version of the validation. Just to ensure our assumptions about YAML/JSON umarshalling are correct and all the cases behave as expected

valaparthvi commented:
> > @valaparthvi @killianmuldoon I've updated the PR description to reflect what we want to achieve. Am I correct?
> 
> Would be really great if someone can test this with a real Cluster/ClusterClass once we have the final version of the validation. Just to ensure our assumptions about YAML/JSON umarshalling are correct and all the cases behave as expected

ykakarap reviewed (commented):
I did some tests locally and the selector is working as as desired.

LGTM pending squash.

killianmuldoon commented:
/lgtm

This is working as intended - thanks @valaparthvi for following this one through!

Could you squash your commits?

valaparthvi commented:
> /lgtm
> 
> This is working as intended - thanks @valaparthvi for following this one through!

Thank you all for quick reviews and patience :)

> Could you squash your commits?

Done

ykakarap reviewed (approved):
/lgtm

fabriziopandini commented:
Getting on this after PTO
@vincepri could you take a look at the API (how do we express the patch applies to all MachineDeploymentClasses)

valaparthvi commented:
@fabriziopandini Do we have any update on this?

vincepri commented on api/v1beta1/clusterclass_types.go:
```diff
@@ -392,6 +392,7 @@ type PatchSelectorMatch struct {
 
 	// MachineDeploymentClass selects templates referenced in specific MachineDeploymentClasses in
 	// .spec.workers.machineDeployments.
+	// Note: If set to an empty object it will match all MachineDeploymentClass.
```
Not sure if we should do this, it's pretty easy for users to match all classes without knowing it this way; have we considered alternative solutions like matching on name globs or labels?

vincepri replied:
/hold

vincepri replied:
To clarify, the glob match should follow https://go.dev/src/path/filepath/match.go

fabriziopandini replied:
I'm also in favor of being explicit in the API and enabling usage of glob inside names allows to do things like `names: [ "*"]` or `names: [ "prod-*"]`

sbueringer replied:
Yes, we had exactly this discussion a while back ~ here: https://github.com/kubernetes-sigs/cluster-api/pull/6930#issuecomment-1187187737

Essentially it comes down to:
* intuitive and easily understandable YAML (with using glob) + somewhat consistency with RBAC (but it only supports "*" afaik) vs.
* consistency with "match all" in other Kubernetes selectors, e.g. NetworkPolicies

The PR was implemented as is because we chose consistency. But I'm fine with reconsidering.

fabriziopandini replied:
I'm sorry if we get late to this discussion, but API design and UX are important and require time and broader consensus, and if I'm not wrong, in this case the preferred option is towards intuitive and easily understandable YAML (with using glob)

sbueringer replied:
Fine for me

vincepri replied:
> consistency with "match all" in other Kubernetes selectors, e.g. NetworkPolicies

While this is an existing API in the Kubernetes space, it also has a lot of drawbacks when applied to other entities. It's not uncommon for the Cluster API project to different in a few places from what we've seen in other places, in most cases it's when it comes down to explicit behavior or security concerns.

I'm in favor of considering adding support for globs in the current names list.

sbueringer replied:
+1

valaparthvi replied:
I have updated the description https://github.com/kubernetes-sigs/cluster-api/pull/6930#issue-1305732557 with acceptable matches. Can someone confirm if it is true before I start working on the changes?

sbueringer replied:
Looks good I'll answer in more detail below

valaparthvi replied:
> I'm also in favor of being explicit in the API and enabling usage of glob inside names allows to do things like `names: [ "*"]` or `names: [ "prod-*"]`

Do you also expect to match `names:["prod-*", "dev-*"]` or just a single element array?

sbueringer commented:
@valaparthvi Sorry for the back and forth. Looks like we don't have consensus yet on how the API looks like. I think that should have happened before we flagged the corresponding issue as help wanted. But we try to improve this in the future.

Please wait with further changes until we have 100% consensus on the API on it's behavior. Just to avoid further back and forth.

sbueringer commented:
@valaparthvi looks like we have consensus now. Sorry for the back and forth. Can you please update the pr accordingly? (i.e. supporting globs in names instead of the current implementation)

Thank you very much!

valaparthvi commented:
> @valaparthvi looks like we have consensus now. Sorry for the back and forth. Can you please update the pr accordingly? (i.e. supporting globs in names instead of the current implementation)
> 
> Thank you very much!

That's alright @sbueringer. I am glad a consensus has been reached.
I have updated the description with acceptable matches. Can you confirm if it is true?

sbueringer commented:
Small clarification:
```
    - selector:
        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
        kind: DockerClusterTemplate
        matchResources: {}
```
=> Should return an error (is already checked today)

```
    - selector:
        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
        kind: DockerClusterTemplate
        matchResources:
          machineDeploymentClass: {}
```
=> Should return an error

```
    - selector:
        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
        kind: DockerClusterTemplate
        matchResources:
          machineDeploymentClass:
            names : [] # empty array
```
=> Should return an error

sbueringer commented:
@valaparthvi 

Small clarification:
```
    - selector:
        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
        kind: DockerClusterTemplate
        matchResources: {}
```
=> Should return an error (is already checked today)

```
    - selector:
        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
        kind: DockerClusterTemplate
        matchResources:
          machineDeploymentClass: {}
```
=> Should return an error

```
    - selector:
        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
        kind: DockerClusterTemplate
        matchResources:
          machineDeploymentClass:
            names : [] # empty array
```
=> Should return an error

valaparthvi commented:
> @valaparthvi Small clarification:
> 
> ```
>     - selector:
>         apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
>         kind: DockerClusterTemplate
>         matchResources: {}
> ```
> 
> => Should return an error (is already checked today)
> 
> ```
>     - selector:
>         apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
>         kind: DockerClusterTemplate
>         matchResources:
>           machineDeploymentClass: {} # I assume names is nil in this case
> ```
> 
> => Should return an error
> 
> ```
>     - selector:
>         apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
>         kind: DockerClusterTemplate
>         matchResources:
>           machineDeploymentClass:
>             names : [] # empty array
> ```
> 
> => Should return an error

I believe all the above cases are already covered today.

sbueringer commented:
Resolved the old & outdated comments.

@valaparthvi Is this PR ready for another round of review or WIP?

valaparthvi commented:
> Resolved the old & outdated comments.
> 
> @valaparthvi Is this PR ready for another round of review or WIP? (just asking to avoid confusion if nobody reviews it because it's marked as WIP :))

I'll give it a round of self review and move it out of WIP. There are a few more changes I need to make to docs.

sbueringer commented:
/retest

sbueringer commented:
I'll review as soon as I get to it, but it will be probably after KubeCon.

sbueringer commented on internal/controllers/topology/cluster/patches/inline/json_patch_generator.go:
```diff
@@ -170,9 +171,15 @@ func matchesSelector(req *runtimehooksv1.GeneratePatchesRequestItem, templateVar
+					if strconv.Quote(mdClass) == strconv.Quote("*") || string(templateMDClassJSON.Raw) == strconv.Quote(mdClass) {
 						return true
 					}
+					regex, err := regexp.Compile(mdClass)
+					if err == nil {
+						if regex.MatchString(string(templateMDClassJSON.Raw)) {
+							return true
+						}
+					}
```
The idea was to use glob matching not regex.

I expected that Go provides a package for that. Apparently there are only filepath.Glob and filepath.Match

Assuming that we don't want to implement glob matching ourselves, did we have any specific library in mind @vincepri @fabriziopandini? (Maybe I'm missing an easy way to implement this).

Btw. I would be absolutely fine with just implementing `*` and not also trying to implement matching for globas in general (like `test-*`)

vincepri replied:
Those are valid concern, I wouldn't expect us to copy the glob package, but we can probably rather start with something simple for the most common use cases. (Solve for the 80% for now)

We should be able to support the following:

- `*` -> everything
- `test-*` -> prefix
- `*-test` -> suffix

A few constraints we should put in place in our own code:

- Every string can at most have a single `*` rune.
- The `*` rune can appear only at the beginning, or ending of the string.
- The rest of the string without the `*` character should comply with Kubernetes naming standards.

sbueringer replied:
Sounds good to me

@valaparthvi We discussed what we consider "comply with Kubernetes naming standards" and it comes down to doing this validation:
```go
			// Replace * with the valid character "a" and then validate if name is a qualified name.
			validation.IsQualifiedName(strings.ReplaceAll(name, "*", "a"))
```

Everything below "A few constraints we should put in place in our own code:" should be added to the `validateSelectors` func in our ClusterClass webhook.

valaparthvi replied:
>The rest of the string without the * character should comply with Kubernetes naming standards.

>We discussed what we consider "comply with Kubernetes naming standards" and it comes down to doing this validation: (+ handle the returned errors)

I am assuming this applies to both strings with/without "*". The first comment mentions without "*", but I think it equally applies to both. Correct me if I am wrong.

valaparthvi replied:
>The idea was to use glob matching not regex.

There is no built-in glob pkg in Golang, would you like me to use an external pkg or can we make do with regex? TBH, I am not sure I understand glob matching. I understand that it is used for file matching, but why do we need it here to match names?

sbueringer replied:
The idea was to have some simple way to match without having to expose full regex support.

The idea now was to do a custom implementation for those rules:
* * -> everything
* test-* -> prefix
* *-test -> suffix

valaparthvi replied:
>The idea now was to do a custom implementation for those rules: (as there is no standard lib for glob)

I'm going to attempt to write https://go.dev/src/path/filepath/match.go for matching name globs.

valaparthvi replied:
WDYT about using an existing pkg: https://github.com/gobwas/glob? This pkg is not actively maintained, but it seems [mature](https://github.com/gobwas/glob#performance) and is used by [thousands](https://github.com/gobwas/glob/network/dependents).

fabriziopandini replied:
Personally, I would avoid a new dependency, especially if not maintained.
Also, what we need is just (StartsWith * and HasSuffix ...)  or (EndsWith * and HasPrefix ...), so it seems overkilling

valaparthvi replied:
> Personally, I would avoid a new dependency, especially if not maintained. Also, what we need is just (StartsWith * and HasSuffix ...) or (EndsWith * and HasPrefix ...), so it seems overkilling

So if you're okay with what we have right now, I can start working on unit tests.

fabriziopandini commented:
If I'm not wrong it is still missing the validation rule suggested in https://github.com/kubernetes-sigs/cluster-api/pull/6930#discussion_r1013081563
Apart from this, I'm +1 to make the final push to get this PR merged as soon as things settle down after 1.3 release is cut later this week

valaparthvi commented:
> 

It's still missing.

> If I'm not wrong it is still missing the validation rule suggested in [#6930 (comment)](https://github.com/kubernetes-sigs/cluster-api/pull/6930#discussion_r1013081563) Apart from this, I'm +1 to make the final push to get this PR merged as soon as things settle down after 1.3 release is cut later this week

I'm sorry I couldn't get to this earlier. I'll get this done ASAP.

valaparthvi commented:
I have updated the code. I'll update the tests once the code has been reviewed.
I'm sorry about the delay. :raised_hands:

fabriziopandini commented:
Validation rules seems ok to me, please add unit tests

valaparthvi commented:
> Validation rules seems ok to me, please add unit tests

https://github.com/kubernetes-sigs/cluster-api/pull/6930#discussion_r1039528341
@fabriziopandini I am kinda waiting for an answer to this.

fabriziopandini commented on internal/controllers/topology/cluster/patches/inline/json_patch_generator.go:
```diff
@@ -170,9 +171,15 @@ func matchesSelector(req *runtimehooksv1.GeneratePatchesRequestItem, templateVar
+					if mdClass == "*" || string(templateMDClassJSON.Raw) == strconv.Quote(mdClass) {
 						return true
 					}
+					regex, err := regexp.Compile(mdClass)
```
can we simply test prefix suffix instead of compiling a regex for every md class?

fabriziopandini reviewed (commented):
I would suggest getting rid of regex and aligning to @vincepri suggestion
* -> everything
test-* -> prefix
*-test -> suffix
The simpler, the better

fabriziopandini commented:
lgtm pending squash
@sbueringer PTAL, this one should be ready to go!

fabriziopandini commented:
it seems there is a test failure to be fixed as weel 😅

sbueringer commented:
Alright, will take a look ASAP

sbueringer commented on internal/webhooks/patch_validation.go:
```diff
@@ -198,6 +199,36 @@ func validateSelectors(selector clusterv1.PatchSelector, class *clusterv1.Cluste
+						"\"*\" rune can only appear at the beginning, or ending of the selector"))
+					break
+				}
+				// a valid selector with/out "*" should comply with Kubernetes naming standards.
```
```suggestion
				// a valid selector without "*" should comply with Kubernetes naming standards.
```
nit

sbueringer commented on internal/webhooks/patch_validation.go:
```diff
@@ -198,6 +199,36 @@ func validateSelectors(selector clusterv1.PatchSelector, class *clusterv1.Cluste
+						"selector does not comply with the Kubernetes naming standards"))
+					break
+				}
+				match = true
+				break
```
```suggestion
```
I think we don't actually check if the selector matches here.

I think the validation here is fine but we should drop match = true and break

sbueringer replied:
I think we have to extend l.233 with the new matching logic (essentially use the same logic as in json_patch_generator instead of just a simple `==`)

sbueringer commented on internal/webhooks/patch_validation.go:
```diff
@@ -198,6 +199,36 @@ func validateSelectors(selector clusterv1.PatchSelector, class *clusterv1.Cluste
+			}
 			for _, md := range class.Spec.Workers.MachineDeployments {
 				if md.Class == name {
 					if selectorMatchTemplate(selector, md.Template.Infrastructure.Ref) ||
```
Can you please modify:
```
            names:
            - default-worker
```

to
```
            names:
            - *-worker
```
in test/e2e/data/infrastructure-docker/v1beta1/main/clusterclass-quick-start.yaml

Just gives us some e2e test coverage that the ClusterClass webhook allows the `*` syntax in selectors

valaparthvi replied:
>in test/e2e/data/infrastructure-docker/v1beta1/main/clusterclass-quick-start.yaml

This file doesn't exist anymore. Can you point me to an alternate one?

sbueringer replied:
Now test/e2e/data/infrastructure-docker/main/clusterclass-quick-start.yaml

sbueringer commented:
/retest

sbueringer commented:
Thank you very much!!

/lgtm

Can you please squash the commits?

sbueringer commented:
Thx!

/lgtm

/assign @fabriziopandini

fabriziopandini commented:
kudos to @valaparthvi for bearing with us for so long! (and sorry for the limited bandwidth in reviewing this work)
that's a nice addition!
/lgtm
/approve

sbueringer commented:
/hold cancel

I absolutely agree. Thank you very much @valaparthvi for the patience and tenacity!! :tada:

valaparthvi commented:
Thank you to you both as well @sbueringer @fabriziopandini and to all those who reviewed this PR. I know I was very slow at times but I appreciate everyone's patience with me :)
